src/nagini_translation/analyzer.py

```diff
@@ -460,15 +460,53 @@ def visit_ClassDef(self, node: ast.ClassDef) -> None:
                 if not (isinstance(base, ast.Name) and
                         base.id in MYPY_SUPERCLASSES):
                     actual_bases.append(base)
+
+        if len(actual_bases) > 0:
+            cls.superclass = self.find_or_create_target_class(actual_bases[0])
+
+            # Process ADT defined by a class hierarchy (sum) and a NamedTuple (product)
+            if isinstance(cls.superclass, PythonClass) and cls.superclass.is_adt:
+                cls.is_adt = True
+                # Detect malformed ADTs
+                # ADT classes can have at most two superclasses
+                if len(actual_bases) > 2:
+                    raise InvalidProgramException(cls.node, 'malformed.adt',
+                            'malformed algebraic data type: superclasses can only ' +
+                            'be a class optionally followed by one NamedTuple')
+                # ADT classes should have no body
+                if len(node.body) != 1 or not isinstance(node.body[0], ast.Pass):
+                    raise InvalidProgramException(cls.node, 'malformed.adt',
+                            'malformed algebraic data type: classes cannot have ' +
+                            'body, fields should be defined in NamedTuples instead')
+                elif len(actual_bases) == 2:
+                    # ADT classes can only define fields by inheriting from NamedTuple
+                    if not (isinstance(actual_bases[1], ast.Call)
+                       and actual_bases[1].func.id == 'NamedTuple'):
+                        raise InvalidProgramException(actual_bases[1], 'malformed.adt',
+                            'malformed algebraic data type: only NamedTuple can be ' +
+                            'used to define fields')
+                    # The name of the NamedTuple should match the ADT being defined
+                    if not (cls.name == actual_bases[1].args[0].s):
+                        raise InvalidProgramException(actual_bases[1], 'malformed.adt',
+                            'malformed algebraic data type: name of NamedTuple has to ' +
+                            'be the same of the class (ADT Constructor)')
+                    
+                    # Parse NamedTuple's fields and their respective types
+                    for field_decl in actual_bases[1].args[1].elts:
+                        field_name, field_type = field_decl.elts
+                        cls.add_field(field_name.s, field_decl,
+                            self.get_target(field_type, self.module))
+
+                    # Consider ADTs as a special case of single inheritance
+                    actual_bases = [actual_bases[0]]
+
```

Comment:
I think we might be missing checks that make sure that the class structure is ADT <- ADT_Class <- ADT_Constructor. As far as I can see, the existing code would allow to declare a constructor for the ADT class itself, or an ADT_Class that has another ADT_Class as a subtype.